spark-platform.js

```diff
@@ -0,0 +1,842 @@
+          }
+          textTextureRenderer._context.fillStyle = textTextureRenderer._settings.textColor;
+          for (let i = 0, n = drawLines.length; i < n; i++) {
+            let drawLine = drawLines[i];
+            textTextureRenderer._context.fillText(drawLine.text, drawLine.x, drawLine.y);
+          }
+
+          if (prevShadowSettings) {
+            textTextureRenderer._context.shadowColor = prevShadowSettings[0];
+            textTextureRenderer._context.shadowOffsetX = prevShadowSettings[1];
+            textTextureRenderer._context.shadowOffsetY = prevShadowSettings[2];
+            textTextureRenderer._context.shadowBlur = prevShadowSettings[3];
+          }
+
+          if (cutSx || cutSy) {
+            textTextureRenderer._context.translate(cutSx, cutSy);
+          }
+          // Original Lightining code ends here
+          canvasInternal.ready.then(() => { // everything is drawn
+            renderInfo.w = canvasInternal.w;
+            renderInfo.h = canvasInternal.h;
+            textTextureRenderer._canvas.width = canvasInternal.w;
+            textTextureRenderer._canvas.height = canvasInternal.h;
+            textTextureRenderer.renderInfo = renderInfo;
+            resolve();
+          });
+        });
+      });
+    });
+    return drawPromise;
+  }
+
+  paint(gl, obj, x, y, color, e) {
+    if (obj) {
+      let translateOnly = true;
+      if (e && e.rotation !== 0) {
+        obj.r = e.rotation * 180/Math.PI;
+        translateOnly = false;
+      } else {
+        obj.r = 0;
+      }
+      gl.beginNativeSparkRendering();
+      if (typeof obj.description === "function" && (obj.description() === "pxWaylandContainer" || obj.description() === "pxSceneContainer")) {
+        obj.paint(x, y, color, false);
+        this.stage.forceRenderUpdate(); // keep updating
+      } else {
+        obj.paint(x, y, color, translateOnly);
+      }
+      gl.endNativeSparkRendering();
+    }
+  }
+
+  getFontSetting(textTextureRenderer) {
+    let fontResource = textTextureRenderer._context.font;
+    let fontFace = textTextureRenderer._settings.fontFace;
+    let fontStyle = textTextureRenderer._settings.fontStyle.toLowerCase();
+
+    if (this._fontResources !== undefined && this._fontResources.has(fontFace)) {
+      fontResource = this._fontResources.get(fontFace);
+      if (fontResource.needsStyleCoercion(fontStyle)) {
+        let url = fontResource.url;
+        fontResource = sparkscene.create({t: "fontResource", url: url, fontStyle: fontStyle});
+      }
+    }
+    return fontResource;
+  }
+
+  addServiceProvider(provider) {
+    if (typeof provider == "function") {
+      sparkscene.addServiceProvider(provider);
+    }
+    else {
+      console.error("service provider need to be a function !!!");
+    }
+  }
+
+  // extensions API
+
+  static createApplicationTexture() {
+    return class ApplicationTexture extends lng.Texture {
```

Comment:
this returns a different class each time

Reply:
@npoltorapavlo true, this was just a quick POC to test if the new approach discussed on Friday worked. Feel free to organise this code in a different way to not instantiate a new class every way.